core/platform/email/gae_email_services_test.py

```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+
+    def test_that_email_not_sent_if_sender_address_is_malformed(self):
+        malformed_sender_email = ''
+        email_exception = (
+            self.assertRaisesRegexp(
+                ValueError, 'Malformed sender email address: %s'
+                % malformed_sender_email))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_mail(
+                malformed_sender_email, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html')
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+
+    def test_that_email_not_sent_if_recipient_address_is_malformed(self):
+        malformed_recipient_email = ''
+        email_exception = (self.assertRaisesRegexp(
+            ValueError, 'Malformed recipient email address: %s'
+            % malformed_recipient_email))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_mail(
+                feconf.ADMIN_EMAIL_ADDRESS, malformed_recipient_email,
+                'subject', 'body', 'html')
+
+
+class BulkEmailsTests(test_utils.GenericTestBase):
+    SENDER_EMAIL = 'sender@example.com'
+    SENDER_USERNAME = 'sender'
+    RECIPIENT_A_EMAIL = 'a@example.com'
+    RECIPIENT_A_USERNAME = 'usera'
+    RECIPIENT_B_EMAIL = 'b@example.com'
+    RECIPIENT_B_USERNAME = 'userb'
+    RECIPIENT_EMAILS = [RECIPIENT_A_EMAIL, RECIPIENT_B_EMAIL]
+
+    def setUp(self):
+        super(BulkEmailsTests, self).setUp()
+        # SENDER is authorised sender.
+        # A and B are recipients.
+        self.signup(self.SENDER_EMAIL, self.SENDER_USERNAME)
+        self.signup(self.RECIPIENT_A_EMAIL, self.RECIPIENT_A_USERNAME)
+        self.signup(self.RECIPIENT_B_EMAIL, self.RECIPIENT_B_USERNAME)
+        self.recipient_emails = [self.RECIPIENT_A_EMAIL, self.RECIPIENT_B_EMAIL]
+
+    def test_that_correct_bulk_emails_sent(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+        messages_a = self.mail_stub.get_sent_messages(to=self.RECIPIENT_A_EMAIL)
+        self.assertEqual(len(messages_a), 1)
+        messages_b = self.mail_stub.get_sent_messages(to=self.RECIPIENT_B_EMAIL)
+        self.assertEqual(len(messages_b), 1)
+
+    def test_that_email_not_sent_if_can_send_emails_is_false(self):
+        # Emails are not sent if the CAN_SEND_EMAILS setting is not turned on.
+        email_exception = (self.assertRaisesRegexp(
+            Exception, 'This app cannot send emails.'))
+        with email_exception:
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.recipient_emails,
+                'subject', 'body', 'html')
+        messages_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_A_EMAIL)
```

Comment:
Can we move this to line above? (Max 80 char in a line)

Reply:
I can do this, but another suggestion of yours prevents me from doing it.

> I think `to=self.RECIPIENT_EMAILS[0]` will make it more readable as the emails are sent to the `self.RECIPIENT_EMAILS`.

Reply:
Thanks for checking!